src/workloads.jl

```diff
@@ -11,30 +12,12 @@ function workload_enabled(mod::Module)
     end
 end
 
-"""
-    check_edges(node)
-
-Recursively ensure that all callees of `node` are precompiled. This is (rarely) necessary
-because sometimes there is no backedge from callee to caller (xref https://github.com/JuliaLang/julia/issues/49617),
-and `staticdata.c` relies on the backedge to trace back to a MethodInstance that is tagged `mi.precompiled`.
-"""
-function check_edges(node)
-    parentmi = node.mi_info.mi
-    for child in node.children
-        childmi = child.mi_info.mi
-        if !(isdefined(childmi, :backedges) && parentmi ∈ childmi.backedges)
-            precompile_mi(childmi)
-        end
-        check_edges(child)
-    end
-end
-
-function precompile_roots(roots)
-    @assert have_inference_tracking
-    for child in roots
-        precompile_mi(child.mi_info.mi)
-        check_edges(child)
+function precompile_newly_inferred(cis)
+    while !isempty(cis)
+        ci = pop!(cis)
+        precompile_mi(ci.def)
     end
```

Comment:
Since this overwrites the internal array which is already enabled by default to collect output for serialization (#43990), this, ironically, may reduce the likelihood that code run during `ex` will be included in the output. 

```suggestion
@noinline function precompile_newly_inferred(cis)
        ccall(:jl_set_newly_inferred, Cvoid, (Any,), nothing) # ensure we drop internally references to both newly_inferred arrays so that it is safe to access them here
        for child in cis # set the must-precompile bit on all new code
            precompile_mi((child::Base.CodeInstance).def)
        end
        append!(Base.newly_inferred, cis) # request the new code be included in the serialized file
        empty!(newly_inferred)
        nothing
end
``````

Reply:
I had forgotten that might trigger additional precompilation, good catch.